README.md
![Starbackjs](https://user-images.githubusercontent.com/45036724/144230732-9104dc0e-c4fe-4c5c-9990-6693f4c0400c.gif)
![Starbackjs Demo Night](https://user-images.githubusercontent.com/45036724/144230828-076a39ae-cc2a-4096-803d-a9bbb49b1d58.gif)

Click for [more examples](https://zuramai.github.io/starback.js/)

## Getting Started
You can use starback.js directly with `<script>` tag or import syntax.
- With `<script>` tag

  ```html
  # Using CDN
  <script src="https://unpkg.com/starback@2.0.1/dist/starback.js"></script>

  # Using downloaded files
  <script src="PATH_TO_YOUR_DIST_FOLDER/starback.js"></script>
  ```

- With import

  Install via NPM first:

  ```bash
  npm install starback
  ```

  and then

  ```js
  import Starback from 'starback'
  ```

## Example Usage

```html
<canvas id="canvas"></canvas>

<script src="PATH_TO_DIST_FOLDER/starback.js"></script>
<script>
  const canvas = document.getElementById('canvas')
  const starback = new Starback(canvas, {
      type: 'dot',
      quantity: 100,
      direction: 225,
      backgroundColor: ['#0e1118', '#232b3e'],
      randomOpacity: true,
  })
</script>
```

You can check the more options in [Options section](#options)

## Options


| Key  | Type   | Description          | Required | Default Value|
| --------------- | ----------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | ----------------------- |
| type | String | **Required.** Star type to use. The values are 'dot' or 'line' | true |    |
| width| Number | Canvas width to set  | no | 800    |
| height    | Number | Canvas height to set | no | 600    |
| backgroundColor | String&#124;Array | Color of the background, use string for solid color, or array of colors for linear gradient | no | #ccc   |
| quantity        | Boolean | Number of stars to show | 100 | false        |
| showFps         | Boolean | Show FPS on the top left screen | no       | false        |
| speed | Number | Star speed | no       | 0.5          |
| starSize        | Number&#124;Array | Use number for fixed star size. Use [minSize, maxSize] value for random the size within the range | no       | [0, 3] for `type: dot`, 100 for `type: line`          |

## Additional options for `type: 'dot'`
| Key  | Type   | Description          | Required | Default Value |
| ---- | ------ | -------------------- | -------- | ------------- |
| direction   | Number | Use 360 degree to set the direction | no | 225  |
| randomOpacity   | Boolean&#124;Array | If `true`, random star will get random opacity. Use array [minOpacity, maxOpacity] to limit the random opacity  | no | false  |
| starColor       | String | Color of the stars, you can use `rgb` or `hex` like css color. | no       | white |

## Additional options for `type: 'line'`
| Key  | Type   | Description          | Required | Default Value |
| ---- | ------ | -------------------- | -------- | ------------- |
| directionY      | Number | The direction the star to move vertically (1 = to bottom, 0 = to top)       | no       | 1 |
| directionX      | Number | The direction the star to move vertically (1 = to right, 0 = to left)       | no       | 1 |
| distanceX       | Number | Distance horizontally the star should reaching at     | no       | 0.1          |
| frequency       | Number | The frequency of star will be rotated      | no       | 10|
| slope | Object | The quadraticCurve coordinate that will be applied to stars. Read [MDN](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/quadraticCurveTo) | no       | {x: 1, y: 1} |
| starColor       | String&#124;Array | Color of the stars, use string for solid color, or array of colors for linear gradient | no       | ['#fb00ff', '#00dde0'] |
| spread          | Number | The spread level of the stars   | no       | 1 |


## Methods

| Name   | Parameter         | Description      |
| ---------------------------- | --------------------------------------- | ------------------------------------------------- |
| `generateStar(amount)`       | amount | The quantity of stars that will be shown          |
| `addToFront(callback(ctx))`  | function(ctx: CanvasRenderingContext2D) | Add to front of the falling star by given context |
| `addToBehind(callback(ctx))` | function(ctx: CanvasRenderingContext2D) | Add to back of the falling star by given context  |

Example method usage: adding image to front

```js
let starback = new Starback(canvas, options)

let mountainImage = new Image()
mountainImage.src = 'images/mountain.png'

mountainImage.onload = () => {
  starback.addToFront((ctx) => {
    ctx.drawImage(
      mountain,
      0,
      0,
      mountain.width,
      mountain.height,
      0,
      canvas.height - mountain.height,
      canvas.width,
      mountain.height
    )
  })
}
```

## Contributing

You can contribute to this repository. See [CONTRIBUTING.md](https://github.com/zuramai/starback.js/blob/main/CONTRIBUTING.md)

## Donate

If you have used this library and it's useful for you, please consider to donate:

[Ko-fi](https://ko-fi.com/saugi) | [Trakteer](https://trakteer.id/saugi)

## License

This library is under MIT license.


CHANGELOG.md
# V2 Release
- All configuration object is moved to `Starback.config`

src/starback.js
import Dot from "./types/dot"
import Line from "./types/line"

/**
 * Default Config
 * @type {Object}
 */
export const StarbackDefaultConfig = {
  width: 800,
  height: 600,
  
  randomOpacity: true,
  showFps: false,
  type: 'dot'
}

/**
 * Starback class wrapper
 * @class Starback
 */
export default class Starback {
  static DefaultConfig = StarbackDefaultConfig

  config = {}

  /**
   * Stores stars' class
   * @type {Dot|Line}
   */
  stars = null

  starTypes = {
    'dot': Dot,
    'line': Line
  }

  /**
   * Starback library
   * @param {HTMLElement|string} Canvas element or the selector
   * @param {Object} options
   */
  constructor(canvas, config = {}) {
    this.canvas = canvas instanceof HTMLElement ? canvas : document.querySelector(canvas)
    /** @type {CanvasRenderingContext2D} */
    this.ctx = this.canvas.getContext('2d')


    // merge config
    this.mergeConfig(config)

    //
    this.repeat = 0

    // storing callbacks
    this.frontCallbacks = []
    this.behindCallbacks = []

    // for calculating fps
    this.fps = 0
    this.lastCalledTime = 0

    // time tracking
    this.lastGenerated = 0

    this.init()
  }
  /**
   * Merge Config
   * @param  {StarbackDefaultConfig|object} instanceConfig
   */
  mergeConfig(instanceConfig) {
    // merge config
    let config = Object.assign(StarbackDefaultConfig, instanceConfig)
    
    // apply config
    this.config = config
    // this.width = config.width
    // this.height = config.height
    // this.speed = config.speed
    // this.direction = config.direction
    // // this.directionY = config.directionY
    // // this.directionX = config.directionX
    // this.starColor = config.starColor
    // this.maxStar = config.maxStar
    // this.slope = config.slope
    // this.starSize = config.starSize
    // this.showFps = config.showFps
    // this.config.backgroundColor = config.backgroundColor
    // this.distanceX = config.distanceX
    // this.frequency = config.frequency
    // this.randomOpacity = config.randomOpacity
    // this.spread = config.spread
    // this.type = config.type
  }

  /**
   * Initialize canvas before render
   */
  init() {
    this.canvas.setAttribute('width', this.config.width)
    this.canvas.setAttribute('height', this.config.height)
    this.stars = new this.starTypes[this.config.type](canvas, this.config)

    this.config = Object.assign(this.stars.defaultConfig, this.config)
    this.stars.config = this.config

    this.generateStar()

    requestAnimationFrame((t) => this.render(t))
  }
  
  
  /**
   * Set background for the whole canvas
   */
  setBackground() {
    let bg

    if (typeof this.config.backgroundColor == 'string') bg = this.config.backgroundColor
    else if (typeof this.config.backgroundColor == 'object') {
      bg = this.ctx.createLinearGradient(this.canvas.width / 2, 0, this.canvas.width / 2, this.canvas.height)

      this.config.backgroundColor.forEach((bgString, index) => {
        bg.addColorStop(index / this.config.backgroundColor.length, bgString)
      })
    }
    this.ctx.fillStyle = bg
    this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height)
  }

  /**
   * Draw the frame into the canvas
   */
  draw() {
    this.behindCallbacks.forEach(cb => cb(this.ctx))
    this.stars.draw()
    this.frontCallbacks.forEach(cb => cb(this.ctx))

    // Show FPS if config.showFps is enabled
    if (this.config.showFps) this.drawFps()
  }

  /**
   * Update everything in the canvas frame including stars
   */
  update() {
    this.stars.update()
  }

  /**
   * Add an object in front of the stars
   * @param {Function} cb Callback function
   */
  addToFront(cb) {
    this.frontCallbacks.push(cb)
  }

  /**
   * Add an object behind the stars
   * @param {Function} cb Callback function
   */
  addToBehind(cb) {
    this.behindCallbacks.push(cb)
  }

  /**
   * The total quantity of stars in canvas
   * @param {Number} amount The number of stars
   */
  generateStar() {
    this.stars.generate(this.config.quantity)
  }

  /**
   * Draw the FPS in the canvas.
   */
  drawFps() {
    this.ctx.fillStyle = 'white'
    console.log(this.fps)
    this.ctx.fillText(`${this.fps} fps`, 10, 10)
  }


  /**
   * Canvas render function
   * @param {DOMHighResTimeStamp} timestamp 
   */
  render(timestamp) {
    if (!this.lastCalledTime) this.lastCalledTime = timestamp

    let deltaTime = timestamp - this.lastCalledTime
    this.fps = Math.round(1000 / deltaTime)
    this.lastCalledTime = timestamp

    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height)
    this.setBackground()
    this.draw()
    this.update()

    requestAnimationFrame((t) => this.render(t))
  }

  
}
